### ISSUE TITLE
Run configlet fmt on all tracks

### ISSUE THREAD

**Issue by kytrinyx** [exercism/meta#95]
In the coming months we will be submitting scripted pull requests to the tracks with various nextercism-related changes to the config files.

By normalizing the config files, we will reduce the amount of noise in those PRs.
See documentation about configlet here: https://github.com/exercism/docs/blob/master/language-tracks/configuration/configlet.md

**Comment by Insti** [comment]
I have several issues with the normalization format used here.

**Comment by Insti** [comment]
While these changes which seem to be the result of sorting keys alphabetically might be easy enough for computers, these files need to be maintained by humans so a more human-friendly organisation format needs to be used.

A consistent cross-track format can still be achieved while maintaining the human friendly formatting but this is not it.

See comments on the changes to the Ruby track config.json: https://github.com/exercism/ruby/pull/756/files

**Comment by NobbZ** [comment]
I do have my issues as well...

1. It is the second round of normalisation
2. UUID is the last attribute of each object, it should be first, since it identifies each exercise, after that we should have the name/slug of the exercise since it identifies the exercise for the human. All the other metadata should follow, with the topics list beeing the last, since it might clutter the object if it is anywere else and has many entries.
3. I prefer to write the empty topics as `[]` instead of `null`, at least under some circumstances, this should be left untouched by the formatter. `[]` means no topics apply, `null` means I haven't looked at it yet. At least thats how I read it.
4. In the maintainers file, github username and given name should be more on top of the object, instead of midway through. Then alumnus and show on website, then the other stuff and maybe the bio last as it is the longest value.

Both files are used, read, and modified by humans, so we should make sure they find their way through.

**Comment by NobbZ** [comment]
Also @petertseng made some good points in exercism/haskell#610, that I haven't realized when I glanced over the erlang PR, because I was to focused on the actual exercises and forgot that there is much more in the file.

**Comment by stkent** [comment]
I merged PRs for the tracks I help maintain, but I also agree with the suggested improvements to later normalizations above.

**Comment by N-Parsons** [comment]
I also agree with the above statements. I think it would be best to make the `config.json` as human-readable as possible to make it easier for contributors to understand.

Out of interest, how much difference would having the keys ordered by importance make compared to having them in the proposed alphabetical order for the planned scripted changes? If it's necessary, could it be done as a temporary change that is reverted to a more human-friendly ordering later?

**Comment by NobbZ** [comment]
Scripting changes basically means, importing the JSON into an internal representation of the script, changing stuff in memory, writing back the representation into the JSON file.

There is no normalisation necessary to do this, but since some JSON encoders do not care for pretty printing or key-order (because it is semantically irrelevant) its not the best thing to rely on the output.

Therefore on any scripted (or manually) change, the formatter is called independently.

This is the only way to enforce a canonical format, which I really had prefered to be discussed before hand, instead of creating additional noise after we had a normalisation about 4 months ago to avoid the noise that is now created.

And also in the former normalisation I do remember discussion about how to pretty print empty arrays, but there its been inline vs. three-line...

**Comment by NobbZ** [comment]
I'm realizing right now that empty lines at the end of the file are normalized away. I prefer to keep them, it's easy to miss the last line on cat if it is missing.

**Comment by kytrinyx** [comment]
I have a few thoughts, but they're still fairly nebulous.

I'm not trying to force you all to do extra work, I'm trying to find a way to make it possible to do the work that I need to do with as little negative effect as possible.

I appreciate the constructive suggestions for how the serializer should work, with the reasoning for why.

> Scripting changes basically means, importing the JSON into an internal representation of the script, changing stuff in memory, writing back the representation into the JSON file.
> 
> There is no normalisation necessary to do this

@NobbZ It sounds like you're saying that you'd prefer that I submit pull requests that make seemingly arbitrary changes along with the actual changes that I need to make. That doesn't seem like something you would say, so I'm pretty sure I'm misunderstanding. Would you mind clarifying?

> It is the second round of normalisation

I expect there to be several more rounds of normalization, as we learn more and make better decisions and improve our tooling.

The first round of normalization was the result of a script that I wrote, and we didn't have a tool at that time. The second is based on a tool that all maintainers have access to.

While I absolutely understand that it would have been ideal to get all the details of normalization right the first time, that hasn't been possible. I didn't have all the information I needed to make a better decision.

Let's get the tool right, and then run normalization across the board. I don't care very hard about the details of the normalization, I do care strongly that we have tooling that lets us script changes, because in the next few months there will be many changes as we work to get v2 right.

**Comment by petertseng** [comment]
I'm responding to a question nobody asked yet not because I think someone will ask it of me (I don't think someone will; the question is a bit too hostile), but rather because my response to the hypothetical question explains my behaviour in the past and now:

> Why didn't you raise all your concerns when we were discussing in https://github.com/exercism/discussions/issues/177 ? If you didn't complain back then, you have no right to complain now.

Given my (everyone's!) limited time, at the time it made sense to only comment on issues that would affect me. I looked at the issue in question and thought "oh, formatting, don't need it, don't plan to run it on my tracks, so don't need to comment at all". I guess that ended when it is proposed to be run on the tracks.

It's already known that I defend the humans, and I don't think I contribute to the conversation by rehashing why.

Here are a few possible ways I can see proceedings from this point. All possible ways should include a possible way to achieve the objective of scripting a change to config.json:

1. We get agreement on an ordering of keys that would be acceptable (on the top level, `language` first, some other things, `exercises` last, etc.)
    * The normaliser could emit the keys in that order rather than alphabetised. For example, if you suppose it is written in Go, https://play.golang.org/p/pJ0hFHfVUq.
    * I understand that that necessitates a new release whenever we change the set of expected keys. If that doesn't jive well with configlet's release model, perhaps consider decoupling.
    * To script a change, add a new key to the struct, and then the marshaller will put it in the right place.
2. We decide that the normaliser will only normalise indentation and will leave all key orderings intact.
    * Now, since indentation is normalised and I therefore know that all keys on the top-level object will have an indentation prefix of (say) 2 spaces, if I wish to add, say, a `track_id` to all config files after `language`, I simply look for the line that has `  "language": "whatever",` and add a line immediately after it that says `  "track_id": "whatever",` at the same level of indentation.
    * But this gets tricky if the resulting line should or should not have a comma, and/or the previous line now needs a trailing comma added. I understand it's a fragile solution, but it is one that is in reach if we do not reach agreement on an ordering.
3. Your suggestion here.

**Comment by petertseng** [comment]
EBWODP:

>  I simply look for the line that has `  "language": "whatever",` and add a line immediately after it that says `  "track_id": "whatever",`

Thanks GFM for eating my initial spaces. The string ``` `  a` ``` renders as ` a`. Instead, I'll just edit my sentence to say "at the same level of indentation".

**Comment by kytrinyx** [comment]
> Why didn't you raise all your concerns when we were discussing in exercism/discussions#177 ? If you didn't complain back then, you have no right to complain now.

I disagree with this hypothetical question, for what it's worth.

Also, I defend the humans, I just don't always understand them. I really want to get this right so that we can (a) minimize work for maintainers, and (b) support scripted changes. And make the humans happy.

**Comment by petertseng** [comment]
This entire comment does not need to be taken into consideration when considering the course of action.

~~Looks like there will be trouble if the normaliser does not act the exact same way as whatever is doing the scripting. If we suppose that the scripting is done in Ruby:~~

```diff
$ ruby -rjson -e 'puts JSON.pretty_generate(JSON.parse(File.read("config.json")))' | diff -u - config.json
--- -	2017-10-24 03:15:18.000000000 +0000
+++ config.json	2017-10-24 03:13:21.000000000 +0000
@@ -786,10 +786,8 @@
       "uuid": "d5997b60-e54c-4caa-beae-8614f0da3fb3"
     }
   ],
-  "foregone": [
-
-  ],
+  "foregone": [],
   "ignore_pattern": "example",
   "language": "Haskell",
   "solution_pattern": "example.*[.]hs"
-}
+}
\ No newline at end of file
```

~~This may affect the choice of technology for the normaliser/future scripts~~

Above argument is invalid. Just run the normaliser after running the script, and that way a minimal diff is achieved

**Comment by kytrinyx** [comment]
> Just run the normaliser after running the script, and that way a minimal diff is achieved

Yepp—that's what I've been doing.

**Comment by bdw429s** [comment]
FWIW, I think this all feels slightly round-peg-in-a-square-hole-ish because we're 
1. Slightly abusing a source control management system to act as a database mangement system
2. Slightly abusing textual diff  tools to act as complex data comparison tools
3. Slightly abusing a data format in which extracurricular whitespace has no meaning to make it have meaning
4. Slightly abusing the unordered nature of JSON objects by wishing they had order for our convenience

There's no good way around # 1 and # 2.  It might be better if we made each exercise's data an atomic file in source control instead of one big file, but that would really just be pushing some technical debt around with diminishing returns.  

I sort of feel that # 4 is wrong of us.  If we wanted order to matter, we should have used an array.  If you want a better way to visualize your data, then build a tool to do so IMO.  I'm new to this, but thus far I really haven't cared where the keys were at in the CFML repo.  I'm not changing or even looking at most of those values on a regular basis, and it's not like I forget what track I'm in so I have to scan the JSON file to remind myself.

So # 3 is the crux of the issue IMO.  The impedance mismatch being that two JSON strings with different whitespace can contain the same effective data, yet we're using a storage system that cares about whitespace.  For the most part, I think we can get away with what we need to accomplish if we:

* Decide on and stick to some basic whitespace rules (which we have...) and make sure they're applied to any pulls.
* Let track owners put the keys in whatever place they want
* Upon formatting, keep the keys in the same place and only affect the whitespace. (The CFML track has it's own JSON formatting routine written in CFML and it does just that.)
* Be prepared for a nominal amount of annoyance at the fact that pretty much only one pull can cleanly touch the `config.json` file at a time without some sort of conflicts.
* All be willing to sacrifice a bit of our OCD since we are, after all, slightly abusing a generic text-based data storage mechanism to be simultaneously whitespace aware, ordered to our liking, easy to diff/merge/read, and take over the world all at the same time :)

**Comment by NobbZ** [comment]
> @NobbZ It sounds like you're saying that you'd prefer that I submit pull requests that make seemingly arbitrary changes along with the actual changes that I need to make. That doesn't seem like something you would say, so I'm pretty sure I'm misunderstanding. Would you mind clarifying?

The paragraph you quote is as explanation to @N-Parsons and just meant as an explanation about the workflow with the normalizer and scripted changes.

Of course I do expect the scripted change and the re-normalisation to happen in a single commit.

---

> I expect there to be several more rounds of normalization, as we learn more and make better decisions and improve our tooling.

Ah, Okay, I can live with that, but it would have been nice to tell that from the very beginning.

---

> Slightly abusing the unordered nature of JSON objects by wishing they had order for our convenience

As a maintainer I have to alter the file by hand if things change, eg. adding a new topic to an exercise or changing its difficulty. Therefore its necessary to be able to find those entries quickly. So when I do a text search for `accumulate`, because thats the exercise I want to change, then after finding it, I do only want to scan the lines below to find the key I want to alter. Having to scan in both directions makes it even harder, especially given the chance that the potentially huge list of topics could be between the exercises name and the key I'm actually searching for.

If we had a graphical tooling[1] though, which were guaranteed to work on Mac, Win and Linux at least that takes care of this and I could alter entries as I'm used to it with any CRUD-tool, I were fine with any ordering.

---

>  only one pull can cleanly touch the config.json file at a time

Personally this is a no-go for me. If scripted changes come in fast, and I'm burried under day jobs work, I were probably unable to handle those quick enough, especially since solving those involves not only me to merge, but also the author of the PR to rebase. Since those PRs will probably done from in timezones raching from UTC-4 to -9, while I do live in UTC+1/+2 (depending on DST). This difference of 5 to 12 hours results often in roundtrips of a complete day.

---

[1] Some go program which serves a small web app is totally fine for me.

**Comment by tleen** [comment]
* I like the idea of cross-track normalization and `configlet fmt` should be the authority on it.
* In Javascript we can't guarantee ordering in JSON so I have become used to not caring much about ordering of keys. Because configlet is in Go we can actually enforce an output key order (yay Go :smile:) so we can actually consider it.

* No strong opinions on output order except it should have some logic to it. Alphabetical is always a safe default, but as has been pointed out here, because `config.json` is hand-edited it may make sense to group keys logically to make that process easier. The Go track's is in alphabetical order and I can't say I even noticed much.

* Targeted PRs make sense to me: "single purpose changes". `nextercism` is a rolling target, it's to be expected that `config.json` will have to adapt more than once.

> I prefer to write the empty topics as [] instead of null, at least under some circumstances, this should be left untouched by the formatter. [] means no topics apply, null means I haven't looked at it yet. At least thats how I read it.

I can see the logic behind this thinking, but it does give us two different representations of "no topics": "no topics because there are none" and "no topics because none have been assigned yet", not sure if we need that distinction. Although if there is an exercise with no applicable topics perhaps we need to expand the topics list or remove the exercise? Even "hello world" has applicable topics.

(I guess technically "I have not defined it yet" would be `undefined` in JSON while `null` would be "I am not defining it"? JavaScript! shine on you crazy :large_blue_diamond:)

Sometimes the programmer in me sees a debate like this and thinks "Can we make everyone happy?" add a switch to have `fmt` output config as alphabetical or maintainer(ical?). But, hopefully we can agree that it is [probably better](https://xkcd.com/927/) to figure out a canonical ordering and stick to it. Otherwise I fear we would just be kicking the can down the road to deal with later, when configs are larger and more complex.

This is a good discussion. Glad we are having it. Nextercism will come out better as a result!

**Comment by marnen** [comment]
I'm sort of jumping into the middle of this discussion since @kytrinyx requested some input from me, so if I'm saying something silly due to context that I'm not aware of, please tell me.

@NobbZ:
> Of course I do expect the scripted change and the re-normalisation to happen in a single commit.

I would very much want two separate commits (possibly in one pull request). Although it's not a hard and fast rule, when code is reformatted without changing content, it's nice for that reformatting to be in a commit of its own, separate from any semantic changes.

***

I am, in general, extremely skeptical of automatic code formatters (except where they're really a core part of a language ecosystem, as I suppose gofmt is), and I have no particular interest in cross-track normalization or stylistic consistency for its own sake. In the config files that I've written, I use the order and style that I personally find most readable, and I'd get pretty cranky if someone told me I had to switch to a style that I considered less readable. Likewise, I'll extend other people that same courtesy for keeping _their_ preferred style. I don't think anything is gained by forcing everyone into the same Procrustean bed. :)

So I guess what I'm saying is that I don't especially want to enforce a canonical order, nor do I see much use for something like `configlet fmt` except as a linter.

I also strongly agree with most of [@bdw429s's comments](https://github.com/exercism/meta/issues/95#issuecomment-338866537): it does seem like we are (slightly) abusing JSON and generally overthinking the concept of normalization.

**Comment by nywilken** [comment]
Its seems like there are a number of issues being discussed in this thread pertaining to configuration file normalization. All of which are valid, but before we start to pick things apart I think it is best to start with a  little context behind `configlet fmt`. 

- The original intent behind its creation was not to enforce a standard order, but to normalize the JSON file in a consistent manner for automated scripting purposes. 
- The reason for the ordered JSON stems from the fact that `configlet` reads in the raw JSON file (it only cares that it is valid JSON) and spits out a formatted JSON file. This was done to reduce the coupling between `configlet` and the two configuration files `config.json` and `maintainers.json`. 
- We needed a way to enforce topic names programmatically across tracks [#60](https://github.com/exercism/configlet/issues/60)

With that said, I'm with @kytrinyx 

>  I really want to get this right so that we can (a) minimize work for maintainers, and (b) support scripted changes. And make the humans happy.

Having a single source of truth for configuration file normalization is key. Achieving this can be simple or tedious, IMO.
- The simplest is changing `configlet` so that it outputs an agreed JSON file format that is easy for maintainers to read. The trade off being that we would need to update `configlet` each time a new field is added, and possibly add some logic to ensure unknown data is not lost while Configlet is being updated. 
- Add some flags to the `fmt` command (e.g `configlet fmt . --normalize-topics-only`). Supporting alphabetical and track-specific ordered configs might not be all that simple 🤔 
- Add support to `configlet fmt` to call out to some other track specific tool for formatting. From our perspective we would just call `configlet fmt .` and rely on the track's tool for normalizing the files as the track maintainers see fit. This is a bit complex, but it has the most flexibility and should make everyone happy 🤞 

For the record, I do agree with @tleen @bdw429s when it comes to the order, but I do understand how this can introduce more work for maintainers working with the config files manually. 

Which brings me @NobbZ's statement about a CRUD like graphical tool. What if `configlet` had support for modifying config data via some CRUD like set of sub-commands (i.e `configlet exercise [add|update|delete]`)? Is that something people would be interested in?

> If we had a graphical tooling[1] though, which were guaranteed to work on Mac, Win and Linux at least that takes care of this and I could alter entries as I'm used to it with any CRUD-tool, I were fine with any ordering.

Regarding PRs

> Targeted PRs make sense to me: "single purpose changes". nextercism is a rolling target, it's to be expected that config.json will have to adapt more than once.

👍

**Comment by bdw429s** [comment]
>  What if configlet had support for modifying config data via some CRUD like set of sub-commands (i.e configlet exercise [add|update|delete])? Is that something people would be interested in?

For what it's worth, I've all but removed the need for track contributors to even need `configlet` in the CFML track.  We've already written CFML-based CLI tools that scaffold exercises (which includes the additions to the `config.json` file automatically), generate readme files, and normalize the `config.json` to the same specs that `configlet fmt` uses.  Part of this was just to reduce the amount of tools needed to a bare minimum for people to contribute and because I am a man with a CFML hammer and the world is my nail... :hammer:  :smile:

**Comment by nywilken** [comment]
@bdw429s good show - I have to take a look at what you've done as I can see value in providing such tools.

**Comment by marnen** [comment]
@nywilken:

> It's original intent behind [`configlet fmt`'s] creation was not to enforce a stand order, but to normalize the JSON file in a consistent manner for automated scripting purposes.

Why would that need normalization, unless our scripts are acting directly on the JSON as text? (And of course, that's not how it should be done in general: JSON should be parsed into appropriate data structures by the script, and then all manipulation should happen on these data structures, serializing them back to JSON only after that's happened.)

> We needed a way to enforce topic names programmatically across tracks #60

I'll look at that issue...but surely if we need to reinforce that, then plain JSON is the wrong tool, no? (Although I hate to suggest it, perhaps we need a monorepo so everything can refer to the same files?)

> What if `configlet` had support for modifying config data via some CRUD like set of sub-commands 

That might be interesting (like `git config`?), but only if it has a decent UI. Otherwise people will go back to editing the config files directly, and we're right back at square one. 

***

@bdw429s:

> I've all but removed the need for track contributors to even need `configlet` in the CFML track. We've already written CFML-based CLI tools

My hat is off to you from a programming point of view, but from the point of view of Exercism maintenance, this worries me a bit. If we have standard tools meant to be used for all tracks, then we should probably not reimplement them, or else they'll surely go out of sync. 

> I am a man with a CFML hammer and the world is my nail

I'm a former ColdFusion developer myself. Friendly advice from that perspective: get yourself some other hammers as well. :)

**Comment by NobbZ** [comment]
@marnen:
> I would very much want two separate commits (possibly in one pull request).

Nope, since this would defend the sole purpose of normalisation: easier to read diffs.

> serializing them back to JSON only after that's happened.

And this is where anarchy begins if you do not normalize the JSON before hand… Since every tool, every JSON serializer might behave differently.

**Comment by NobbZ** [comment]
@bdw429s:
> [beeing proud of reimplementing configlet in the tracks language]

Why should I put time into recreating a tool that is already there and that is cross platform. This tool is even available precompiled for MacOS X, Linux 32 and 64 bit as well as ARMv5 and v6, Windows 32 and 64 Bit…

It is well tested by the maintainers of all - 1 tracks, known to work as expected, and when specs change, its the first tool who knows, since this tool actually **is** the spec…

The Not-Invented-Here-Syndrom you suffer from though, is what I've often seen to ruin projects, because they got to expensive after months of developer time were spent to rebuild something that was available for only 1k USD/year, thats less than a week of a developers time costs, not even to speak about his desk, PC, powerusage etc…

But we are not here to discuss NIHS. Feel free to stick with your tool, but please do not complain when later on there will be the scripted PRs which will cause a lot of noise on your diffs or when your track can't get processed or served properly because you didn't pick up a small change in the specs of those files fast enough.

**Comment by nywilken** [comment]
@marnen 

> Why would that need normalization, unless our scripts are acting directly on the JSON as text? 

There was a concern that programmatically modifying config files would result in inconsistencies when it came to idententaion, since config files are modified differently across tracks. So the fmt command was built to normalize “format” JSON files (via JSON serialization and deserialization in Go) to ensure that files followed a standard level of indentation and were in a reproducible structure. The fmt command is idempotent, and should always result in the same file structure with little to no changes specific to indentation; making scripted PRs for config files easier to consume and free of major white space changes.  At the current moment, the use of `configlet fmt` is being ran for the first time on some tracks so there are a number of  white space changes in the PRs referencing this issue.  But that should not be the case for future PRs. 

I hope that answers your first question.

**Comment by NobbZ** [comment]
@nywilken:
> Which brings me @NobbZ's statement about a CRUD like graphical tool. What if configlet had support for modifying config data via some CRUD like set of sub-commands (i.e configlet exercise [add|update|delete])? Is that something people would be interested in?

Personally I'm fine with such an approach, I do often prefer CLI over GUI. But we have to ask the other maintainers as well. Especially in the more Windows and Mac oriented languages, I think there will be demand for something more GUI like than an interactive CLI tool.

**Comment by marnen** [comment]
@NobbZ:

> Nope, since this would defend the sole purpose of normalisation: easier to read diffs.

I can't make this make sense, which probably means I'm not understanding you correctly. Can you explain what you meant?

> And this is where anarchy begins if you do not normalize the JSON before hand… 

Quite the contrary. The normalization (if any) happens *on serialization*, so any normalization done on input is completely irrelevant. 

> Since every tool, every JSON serializer might behave differently.

I have 2 responses:

1. There's only one tool in question AFAIK, that being `configlet`. Am I misunderstanding?
2. Even if there were multiple ones, and even if they for some reason normalized their serialized output differently, I wouldn't care much as long as the semantics were the same.

**Comment by NobbZ** [comment]
@marnen:
>>Nope, since this would defend the sole purpose of normalisation: easier to read diffs.
>
>I can't make this make sense, which probably means I'm not understanding you correctly. Can you explain what you meant?

As long as we have a normalized base and re-normalize after each change, resulting diffs will be kept clean. There will be no jumping keys because of different or even random ordering of keys in an object on serialisation.

>>And this is where anarchy begins if you do not normalize the JSON before hand…
>
>Quite the contrary. The normalization (if any) happens on serialization, so any normalization done on input is completely irrelevant.

Well, you need to normalize it once to have a clean base, make tabula rase.

Of course you can decide to not accept those first clean up and do your thing, but the first scripted change including normalisation will come now or later.

> There's only one tool in question AFAIK, that being configlet. Am I misunderstanding?

The scripted changes can be done by tools written in any language by using any JSON library. Therefore `configlet fmt` should be run afterwards to retain normalized format.

> Even if there were multiple ones, and even if they for some reason normalized their serialized output differently, I wouldn't care much as long as the semantics were the same.

But I do.

When a scripted change comes in, which is purpose is to multiply all difficulties by 10 to scale up to the new range from 0 to 100 instead of 0 to 10, and this change does also swap the keys of the object around, I am not able to properly review those changes because the diff gets to noisy.

Therefore we normalize, to be able to keep small focusable diffs.

I do not trust foreign changes, I do not trust foreign scripts, I do not even trust my own, this is why I like to get reviewed, this is why I really miss a second pair of eyes for the erlang track…

**Comment by marnen** [comment]
@NobbZ:

> When a scripted change comes in, which is purpose is to multiply all difficulties by 10 to scale up to the new range from 0 to 100 instead of 0 to 10, and this change does also swap the keys of the object around, I am not able to properly review those changes because the diff gets to noisy.

Right, if formatting and semantic changes are mixed in one commit, then that commit's diff is less clear. This is why, if that happens, I was recommending above that it be done in two commits. 

> Therefore we normalize, to be able to keep small focusable diffs.

Of course diffs should be small and focused. But I believe that the way to get them that way is to use the VCS to make small focused commits. Changing the source code itself solely in order to make the diffs smaller is putting the cart before the horse and making a fetish out of the tools. 

IMHO the source code should take a form that makes *it* readable, not one that makes the diffs so. Likewise, the diffs should take a form that makes *them* more readable, not the source code.

**Comment by marnen** [comment]
@NobbZ (Also just saw your comment about Erlang. I'm not exactly an expert, but I do have a clue about Erlang and I can take a look...)

**Comment by NobbZ** [comment]
> Right, if formatting and semantic changes are mixed in one commit, then that commit's diff is less clear. This is why, if that happens, I was recommending above that it be done in two commits.

So you suggest to start from a normalized base. Run scripted changes which do change data and because the serializer does random order of keys each time, you get a huge diff. You do a first commit. After that you renormalize resulting in another huge diff. 

This results in 2 commits having huge diffs in their own but changing only 2 lines net.

**Comment by marnen** [comment]
No, that's absolutely not what I'm suggesting. I'm suggesting that if normalization needs to happen as part of a script, it should be in its own commit. 

> This results in 2 commits having huge diffs in their own but changing only 2 lines net.

And if that were the case, then it would be easier to review the branch diff than each commit's diff individually. That's fine. The reverse is fine too. Git works for us, not the other way around.

**Comment by marnen** [comment]
...oh, and I'm also suggesting that the script doesn't need to start from normalized input (or indeed to create normalized output, though it will probably do so since serializers tend to behave predictably).

**Comment by marnen** [comment]
...thinking this over some more. I have no particular desire for every script to arbitrarily change the order of JSON properties, but I also don't want to have that order mandated for the config files themselves. Preserve input order on output, perhaps?

**Comment by NobbZ** [comment]
Which you cant. Let's assume javascript. It will randomized keys on each
serialisation even if no objects have actually changed.

Marnen Laibow-Koser <notifications@github.com> schrieb am Mi., 25. Okt.
2017, 01:02:

**Comment by marnen** [comment]
@NobbZ You actually could (by reading key order when parsing, and using `toJSON` to customize the behavior of `JSON.stringify`), but I'm not sure it would be worth it. :) (I suspect Ruby would do better on this, because hash keys are ordered. No idea about any other language's JSON library, because I rarely care.)

**Comment by marnen** [comment]
Anyway, I've probably gone on way too long here, and for that I apologize. I wouldn't be absolutely opposed to normalization if we could guarantee a nice human-readable order of parameters as @petertseng suggested [above](https://github.com/exercism/meta/issues/95#issuecomment-338859005), although I do still sort of think that any approach that cares about the serialization order of an unordered structure is kind of doing it wrong...

**Comment by bdw429s** [comment]
> from the point of view of Exercism maintenance, this worries me a bit. 

@marnen  I understand and have thought through the same thing.  It's been really a trivial amount of work, and to be honest, it's work for me, not for anyone else, and I'm ok with that.

> get yourself some other hammers as well.

@marnen  Heh, yes-- that was being a little tongue in cheek on my part.  Trust me, I have many hammers in my toolbelt.  CFML is personal to me however, and I have a secondary goal in this Excercism track of using it to showcase just how modern the language is.  Heck, most people out there don't even realize you can _do_ CLI scripting in CFML these days!  This has been in part an experiment in what I can accomplish for the sake of learning.  I'm well aware that the onus will be on me to keep any of these CLI scripts up to date, but they've all been very trivial to date thus the nice seamless tie in to the existing CLI exercise generation is worth it IMO.

> Why should I put time into recreating a tool that is already there 

@NobbZ It sounds like _you_ shouldn't.  My reasoning for doing so is above :)

> since this tool actually is the spec

@NobbZ I would disagree. The spec is the spec.  Configlet is the official implementation of that spec.  As a track maintainer, I'm ok with taking on the responsibility of creating an end-to-end set of CLI scripts which not only follow the spec, but also do much more than configlet does.  I'm scripting complete exercise scaffolding, unit test generation, and JSON manipulation.  The fact that I tossed in the readme generation and JSON normalization to the package is just a side note. 

> The Not-Invented-Here-Syndrom you suffer from though, is what I've often seen to ruin projects, because they got to expensive after months of developer time were spent to rebuild something that was available 

@NobbZ I appreciate your concern, and in most business scenarios I would likely agree.  Let's be realistic though, I spent about 5-10 minutes hammering out a nice CFML-based JSON formatter that's all of 13 lines of code.  I'm not running a for-profit here, this is a hobby and one that I specifically want to showcase modern CFML that fits nicely into the existing toolchain a CFML developer will have installed.

>  but please do not complain 

@NobbZ I assure you, I won't be complaining I'm well aware of what I'm doing.  :smile: 

> later on there will be the scripted PRs which will cause a lot of noise on your diffs or when your track can't get processed or served properly because you didn't pick up a small change in the specs of those files fast enough.

@NobbZ Actually, no, that won't happen.  If the spec (and therefore `configlet fmt`) is changed, there will be another round of pulls to update the format of all the configs (much like this one), at which point I'll simply update my tooling to match.  If it ever becomes a burden, I can easily drop my version and simply switch to running configlet.  

***

Back to the discussion, I think most people are more in agreement than not... well maybe, lol.  First and foremost, for this to work at all, I think we all need to agree to whatever formatting/normalization tool we're using, it needs to be run every time any change is made to the JSON, and in my opinion, it should be part of the same commit.  Assuming that the JSON file is proper to begin with, that means that every diff will _only_ show the actual changes that were made to the JSON, which is what we're going for.   

I'm also assuming that any automatic formatting/normalization should never re-order the JSON arbitrarily.  That means that if a track owner wants his/her keys in a specific order, that will never be changed.  

The only remaining loose end is if a track owner wants to automate a change to their JSON using a language/parser that doesn't preserve order of an object's keys AND that same track owner doesn't want their keys changed.  That will be on them to deal with.  So long as any sweeping "changes from above" are made via a lang like Go that happens to preserve the order, there shouldn't be any issues as long as the JSON is normalized before and after the change.  

So, in short, I'm proposing that the spec for the JSON doesn't include specifics on the order of keys, but only the whitespace like indentation, spacing, and line break characters.  Let the track owners decide if they want to order keys, etc and the formatter won't touch them.

**Comment by Stargator** [comment]
I lot of things I'm for I have been mentioned. Though to be honest, I haven't given the config and maintainer files much thought.

I definitely agree that it needs to be as human readable as possible. So that from a glance you can get the information you want.

**Comment by NobbZ** [comment]
> like Go that happens to preserve the order

It doesn't. If the serializer is based on a struct, it will emit the keys in the same order as they are specified in the go sourcecode of the struct, if though the serializer is based on `map[string]interface{}`, keys will be randomized. If you have a `[]map[string]interface{}` (array of objects) order of keys will even differ between each object.

**Comment by petertseng** [comment]
> What if configlet had support for modifying config data via some CRUD like set of sub-commands (i.e `configlet exercise [add|update|delete]`)? Is that something people would be interested in?

I don't think I'd be interested; I think I'd just do it via a text editor. I think the reason is just that I don't want to have to learn these new commands, whereas I already know how to use my text editor. Seems like too much work (both for those who would implement those commands and those who would learn them) for too little value.

I also don't want to be in a world where such commands are necessary (if, for example, config was a binary format that wasn't human-readable at all), but I don't think anyone was suggesting a move to that world

**Comment by petertseng** [comment]
Something I wonder about.

The body of this issue does a good job of explaining "Why?"

> By normalizing the config files, we will reduce the amount of noise in those PRs.

The title of this issue answers the question "What?"

> Run configlet fmt on all tracks

I wonder whether the title is directing the discussion too much toward the merits of one specific "What?" and not enough toward "How do we achieve the goal we are trying to drive toward?". Generally I think we are staying on the right track; I hope it stays that way.

Maybe the way to achieve the goal isn't necessarily via using `fmt` on all config files, I don't know. I don't have any other suggestion right now, but I wouldn't want to prevent anyone from thinking about it.

> I'm proposing that the spec for the JSON doesn't include specifics on the order of keys, but only the whitespace like indentation, spacing, and line break characters. Let the track owners decide if they want to order keys, etc and the formatter won't touch them.

This proposal is one of two that would currently be acceptable to me (the other possibility being to do the same, but also specify key ordering).

**Comment by petertseng** [comment]
Let's hear it if there are any predictions (other than to add track_id) for what changes will be necessary in the future.

Let me list those that have happened in the past:

* Delete `ignored` directory list on the top level: https://github.com/exercism/meta/pull/7
* In the `repository` key on the top level, remove a leading x: https://github.com/exercism/meta/issues/1
* Add `uuid`, `core`, `unlocked_by` **to each exercise**, since they are needed by Nextercism: https://github.com/exercism/meta/issues/16
* Remove `slug` and `repository` on the top level, as they are no longer necessary: https://github.com/exercism/meta/issues/19

> Maybe the way to achieve the goal isn't necessarily via using `fmt` on all config files

I honestly was so tempted to suggest using text transformations (add this key one line above this other key, with the same level of indentation), but there is too much madness here: 

* If the key already exists, don't add it.
* If the new key is in the last position, that line doesn't get a trailing comma; the previous line does.
* The reverse of that rule if the JSON file uses leading commas instead
* Deleting an entire list means you need to delete multiple lines until you see the next closing brace.

Well, it was worth a try.

How about I ask this question instead. I hope this isn't considered a pointless exercise; it is to help understand the problems we're trying to solve.

Explain why it is preferable for one person to personally script these changes and submit them as PRs to all tracks. Compare it to alternatives such as filing an issue against each track and letting the track maintainers do it.

Well, about those PRs that added `uuid`, `core`, `unlocked_by` to each exercise, it would have been a huge duplication of effort to make each track maintainer script that change individually. It makes sense to simply write the code once and do the track maintainers the favour, otherwise it might never get done! Wouldn't want to have tracks that procrastinate too much and hold back Nextercism.

I guess I understand. Now to see how to make that work.

**Comment by petertseng** [comment]
> EBWODP:

It bears noting that `uuid`/`core`/`unlocked_by` acted on each exercise, whereas all other keys acted upon were at top level.

**Comment by bdw429s** [comment]
> It doesn't.

@NobbZ Ah, crap.  Sorry then I misunderstood.  It sounds like it would be necessary to rewrite configlet to not parse the JSON, but simply format it's whitespace as a pure string manipulation.  This is exactly what the CFML lib does that I'm using.  It never deserializes the JSON into native data types, but simply reads the string, iterates over each character, and applies the whitespace rules as defined (indentation, spacing, line breaks, etc).  In that way, it will never modify the order of the data, it will only format the white space around it.

**Comment by Insti** [comment]
My first preference would be to not mess with the track's formatting at all.

Change to an existing key: Change in-place.
Delete an existing key: delete in-place, move everything after up/left.
Add a new key: Put it at the end.

**Comment by NobbZ** [comment]
@Insti Can you name a tool that does this, preserving all order and indentation and stuff?

**Comment by cmccandless** [comment]
@NobbZ @Insti Well, for what it's worth, [I just wrote one](https://gist.github.com/cmccandless/cdd8741fe258ea3d3f3ccc5b97ef538f) in less than an hour. Python's json library maintains key order. This script will even auto-detect existing key order for adding new exercises.

**Comment by Insti** [comment]
> Can you name a tool that does this, preserving all order and indentation and stuff?

No, but:
a) It wouldn't surprise me if several already existed.
b) It would be easy to write one. (see above)

We are paying the price for using JSON for configuration files. It seems our options are:

* Live with randomly ordered data. (undesirable for many humans.)
* Change to a different file format. (seems unlikely.)
* Add some wrapping arrays to ensure JSON treats elements as ordered. 
* Ensure our tools write out their "JSON" in a specified order.
* Something else?

**Comment by tleen** [comment]
It may be time to start looking towards consensus in this issue. Here is what I am reading:

1. **JSON ain't great**
No it is not, and its limitations are well known: can't comment in it, and JavaScript serializers could care less about key order. However, JSON has more-or-less become the default format for configurations across the board (not just in Exercism), so there is no real push to change `config.json` to config.?.
So how to deal with the JSON?

2. **config.json is hand-edited**
For the most part maintainers manually edit the `config.json` file. This file is becoming quite big in some tracks. Track maintainers have ways they like to edit the `config.json` file and scripted updates can be disruptive to that process.
Due to the hand editing, anything that makes it easier for human readability is a Good Thing.

3. **Property order matters**
Due to the point above, the ordering of the properties for the exercises in `config.json` matters to most. For those it does not matter, there is no strong preference beyond it being logically consistent and common across tracks. I don't see anyone advocating for *alphabetical* ordering or anything else, but when there is a preference it is universally for something that is easy for humans to read and understand in the exercise context, i.e. exercise slug on top.

First attempt at a conclusion: seems like we are overall for __defining some ordering/grouping on the properties of the exercises in the config file and have any tooling default to using that ordering__. Anybody strongly against this?

**Comment by petertseng** [comment]
Just curious:

> JSON has more-or-less become the default format for configurations across the board (not just in Exercism)

Got any corroborating sources for this information? I feel curious about how projects do configuration. I tried to ask my preferred search engine about "most popular configuration file format" but all I got was http://www.terminally-incoherent.com/blog/2013/08/28/what-is-your-favorite-congfig-file-format (a poll which may not be representative) and posts advising not to use JSON (these posts may have been necessary because JSON is the most popular, but that's not explicitly stated).

---

These are my preferences (and thus, I gave 👍  to the above proposal):

1. Most preferred: Normaliser changes whitespace only and leaves ordering intact. To script the addition of new keys, place them immediately after or immediately before a key that is related to it.
1. Normaliser changes whitespace and ordering, using an ordering that we discuss and agree on.
1. No normalisation and just open issues against all tracks if we desire a config.json change, and let track maintainers do it on their own time. (How to deal with tracks that miss a hard deadline to complete the change?).
1. Further discussion to find more alternatives. (I'd consider a different file format if y'all cared, but let's see the alternatives and benefits)
1. Normaliser changes whitespace and alphabetises ordering.
1. No normalisation and some poor victim has to attempt to do text processing on the JSON in order to add new keys (I honestly suspect that this will work for the majority of JSON files for simple changes, but it's just too fragile for me to recommend, and hard to make it work on more involved changes)
1. Least preferred: No normalisation and some poor victim has to manually make PRs to each track by hand if we desire a config.json change that should go out to all tracks.

**Comment by tleen** [comment]
> Got any corroborating sources for this information?

Only observational. If [you](https://learn.perl.org/examples/configuration_files.html) [look](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration?tabs=basicconfiguration) [at](https://github.com/Dynalon/JsonConfig)  [configuration](https://github.com/lightbend/config) [managers](https://github.com/spf13/viper) [for](https://github.com/nmische/cf-configmanager) [non-JavaScript](https://github.com/hassankhan/config) [languages](https://github.com/mehcode/config-rs) as well as [other](https://puppet.com/docs/puppet/5.3/man/config.html) [types](https://blog.heroku.com/json_schema_for_heroku_platform_api) [of](https://cloud.google.com/deployment-manager/docs/configuration/templates/using-schemas) [tools](https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-authoring-templates) they at least [support JSON](https://gohugo.io/getting-started/configuration/) (if not prefer it), as well as any web API. There is no universal ranking system of config file formats so seeing what is supported the most may be the best you can do. I don't think there is a lot of disagreement on this point?

Although YAML is popping up more and more...

**Comment by kytrinyx** [comment]
> First attempt at a conclusion: seems like we are overall for defining some ordering/grouping on the properties of the exercises in the config file and have any tooling default to using that ordering. Anybody strongly against this?

This makes a lot of sense to me, however I'm going to leave you all to come to a consensus, since the outcome affects you more than me.

> Let's hear it if there are any predictions (other than to add track_id) for what changes will be necessary in the future.

None of these are certain, but here are some things that have come up.

* Add explicit defaults for all missing keys, to make the config less magical.
* Group, rename, redefine the various regex patterns.
* Normalize topics.
* Restructure the exercise objects to nest the exercises that they unlock.

There may also still be some deprecated fields that need updating. I've lost track.

**Comment by bdw429s** [comment]
>  we are overall for defining some ordering/grouping on the properties of the exercises in the config file and have any tooling default to using that ordering. Anybody strongly against this?

I'm pretty much with @petertseng 's option # 1.

I don't think you're going to get every track to agree on a global order for any given keys in the JSON so I think it's a bad idea to go there, however I think it's been clearly stated that most maintainers will be happy so long as no outside force changes the order as they've carefully crafted it.  Therefore, I think the spec that we "force" on everyone should only apply to white space with the caveat that any ordering in the JSON should be left as-is.  The over-arching goal here really is to make clean diffs in Git, and that just boils to whitespace.  No ordering needs to be enforced to have clean pulls.

The only other item is where do brand new properties go in the JSON and my thought on that is it really doesn't matter.  You won't pick something that makes everyone happy, so just stick them in and the track maintainers can do a one-time fix to move them wherever they see fit.  And, so long as order is always maintained thereafter, they'll never need "fixed" again.  

So in short, my vote is:
* Make and enforce rules on whitespace only to keep clean diffs.  (the ones we have now are fine for me)
* Any time you touch someone else's JSON, don't change the order of anything to maintain individuality/preference
* Put new keys wherever and let the maintainers of that track move them where they wish based on whatever criterion or warm fuzzy they feel like.

**Comment by tleen** [comment]
> I don't think you're going to get every track to agree on a global order for any given keys in the JSON so I think it's a bad idea to go there, however I think it's been clearly stated that most maintainers will be happy so long as no outside force changes the order as they've carefully crafted it.

I do think there is a common thread in ordering we can probably flesh out. We may find that some maintainers **demand** that uuid be the top field and slug after. But I am willing to bet that logical and common groupings will win the day. Then we get a unified config format, but in a  human readable way maintainers like. Best of both worlds!

Brand new properties will be a thing. I have been [an active proponent of this](https://github.com/exercism/go/issues/798), but found that it experimentally did not take :disappointed:. @bdw429s are you using track-specific fields in the CFML config? What are they? This is informative!

Now a lot of this I think depends on Go JSON handling if we are looking at `configlet` as the handler for these issues. As of now I tend to personally avoid Go for JSON-specific stuff (since Node is better), but I will explore this more to see how well we can handle this in (configlet's) Go. I need to explore how undefined keys are handled. From what I understand we define the format and properties are written as such, I still am uncertain on how undefined properties are written. 

I could see exercise properties/keys being defined as:

`identifiers` (uuid, slug)
`language-specific` (difficulty)
`track-relationships` (core, unlocked_by)
`exercism-relationships` (topics) 
`self` (*?)

FWIW, I have a few utilities for understanding all track configurations as a whole:
* [uniconfig](https://github.com/tleen/exercism-uniconfig) - Stick all the track configs into a single JSON file (which you can then [jq](https://github.com/stedolan/jq)).
* [exercism-config-visualizations](https://github.com/tleen/exercism-config-visualizations) - get statistics about configurations across tracks.
* [exercism-config-to-csv](https://github.com/tleen/exercism-config-to-csv) Turn your config file into CSV for analysis.

Initially these utilities were a way to wrap my head around track configuration so we could do it in the Go track. But these tools gave me a good amount of insight into how track configurations work, and across all tracks. It led to some [issues](https://github.com/exercism/problem-specifications/issues/884) and I encourage track maintainers to look at some other tracks to see where we can consolidate changes. 

Please let me know what you think about the potential property ordering. (and tooling if it helps!)

**Comment by petertseng** [comment]
> I don't think you're going to get every track to agree on a global order for any given keys in the JSON

> I do think there is a common thread in ordering we can probably flesh out.

Please back up with data.

```ruby
require 'json'
require 'set'

# path => ordering => tracks
orders = Hash.new { |h, k| h[k] = Hash.new { |hh, kk| hh[kk] = Set.new } }

Dir.glob('tracks/*/config.json') { |fn|
  track = File.basename(File.dirname(fn))

  record = ->(path, h) {
    orders[path][h.keys] << track
    h.each { |k, v|
      subpath = "#{path}#{?/ unless path == ?/}#{k}"
      record[subpath, v] if v.is_a?(Hash)
      v.each { |ve| record[subpath, ve] if ve.is_a?(Hash) } if v.is_a?(Array)
    }
  }

  record[?/, JSON.parse(File.read(fn))]
}

orders.each { |path, orderings|
  puts "* Path #{path}"
  orderings.sort_by { |_, v| -v.size }.each { |k, v|
    puts "  * #{k} - #{v.size} tracks - #{v.to_a}"
  }
}
```

* Path /
  * ["active", "exercises", "foregone", "language"] - 11 tracks - ["racket", "objective-c", "common-lisp", "swift", "cpp", "elixir", "lfe", "ocaml", "scheme", "elisp", "clojure"]
  * ["language", "active", "exercises", "foregone"] - 7 tracks - ["erlang", "scala", "elm", "python", "ceylon", "purescript", "prolog"]
  * ["language", "active", "test_pattern", "exercises", "foregone"] - 4 tracks - ["crystal", "factor", "ecmascript", "coq"]
  * ["active", "exercises", "foregone", "language", "test_pattern"] - 4 tracks - ["plsql", "lua", "dart", "javascript"]
  * ["language", "checklist_issue", "active", "foregone", "exercises"] - 3 tracks - ["nim", "powershell", "vbnet"]
  * ["language", "active", "test_pattern", "foregone", "exercises"] - 3 tracks - ["coffeescript", "mips", "perl5"]
  * ["active", "blurb", "exercises", "foregone", "language"] - 3 tracks - ["fsharp", "r", "csharp"]
  * ["active", "checklist_issue", "exercises", "foregone", "language"] - 2 tracks - ["c", "bash"]
  * ["language", "active", "solution_pattern", "ignore_pattern", "exercises", "foregone"] - 1 tracks - ["haskell"]
  * ["active", "exercises", "foregone", "ignored", "language"] - 1 tracks - ["idris"]
  * ["active", "blurb", "exercises", "foregone", "ignore_pattern", "language", "solution_pattern", "test_pattern"] - 1 tracks - ["cfml"]
  * ["language", "checklist_issue", "active", "test_pattern", "exercises", "foregone"] - 1 tracks - ["d"]
  * ["language", "blurb", "active", "exercises", "foregone"] - 1 tracks - ["rust"]
  * ["blurb", "language", "checklist_issue", "active", "test_pattern", "exercises", "foregone"] - 1 tracks - ["sml"]
  * ["active", "blurb", "exercises", "foregone", "ignore_pattern", "language"] - 1 tracks - ["go"]
  * ["active", "checklist_issue", "exercises", "foregone", "language", "test_pattern"] - 1 tracks - ["haxe"]
  * ["language", "active", "foregone", "exercises"] - 1 tracks - ["php"]
  * ["language", "checklist_issue", "active", "blurb", "test_pattern", "foregone", "exercises"] - 1 tracks - ["perl6"]
  * ["active", "blurb", "exercises", "language"] - 1 tracks - ["delphi"]
  * ["active", "exercises", "foregone", "language", "solution_pattern"] - 1 tracks - ["kotlin"]
  * ["track_id", "language", "active", "test_pattern", "exercises", "deprecated", "foregone"] - 1 tracks - ["gnu-apl"]
  * ["active", "blurb", "checklist_issue", "exercises", "foregone", "language", "test_pattern"] - 1 tracks - ["groovy"]
  * ["active", "blurb", "exercises", "foregone", "language", "test_pattern"] - 1 tracks - ["typescript"]
  * ["active", "blurb", "exercises", "foregone", "language", "solution_pattern"] - 1 tracks - ["java"]
  * ["language", "active", "test_pattern", "exercises", "ignored", "foregone"] - 1 tracks - ["vimscript"]
  * ["language", "active", "test_pattern", "solution_pattern", "exercises", "foregone"] - 1 tracks - ["fortran"]
  * ["language", "active", "blurb", "gitter", "solution_pattern", "exercises", "foregone"] - 1 tracks - ["ruby"]
  * ["language", "active", "exercises"] - 1 tracks - ["pony"]
  * ["language", "active", "blurb", "exercises", "foregone"] - 1 tracks - ["julia"]
* Path /exercises
  * ["core", "difficulty", "slug", "topics", "unlocked_by", "uuid"] - 30 tracks - ["racket", "objective-c", "idris", "cfml", "plsql", "common-lisp", "swift", "cpp", "go", "elixir", "haxe", "c", "delphi", "fsharp", "lfe", "kotlin", "ocaml", "groovy", "lua", "scheme", "ecmascript", "typescript", "dart", "r", "elisp", "clojure", "java", "javascript", "csharp", "bash"]
  * ["uuid", "slug", "core", "unlocked_by", "difficulty", "topics"] - 29 tracks - ["haskell", "erlang", "crystal", "scala", "nim", "elm", "d", "rust", "sml", "coffeescript", "php", "perl6", "factor", "python", "powershell", "mips", "ecmascript", "ceylon", "perl5", "purescript", "coq", "vbnet", "vimscript", "fortran", "javascript", "prolog", "ruby", "pony", "julia"]
  * ["deprecated", "slug", "uuid"] - 8 tracks - ["racket", "swift", "go", "lfe", "kotlin", "ocaml", "java", "javascript"]
  * ["uuid", "slug", "deprecated"] - 8 tracks - ["haskell", "scala", "rust", "sml", "perl6", "python", "perl5", "ruby"]
  * ["core", "difficulty", "slug", "topics", "uuid"] - 6 tracks - ["c", "delphi", "typescript", "r", "javascript", "bash"]
  * ["uuid", "slug", "difficulty", "topics"] - 4 tracks - ["rust", "php", "python", "ruby"]
  * ["difficulty", "slug", "topics", "uuid"] - 2 tracks - ["cfml", "common-lisp"]
  * ["uuid", "slug", "core", "difficulty", "topics"] - 2 tracks - ["gnu-apl", "ecmascript"]
  * ["uuid", "core", "difficulty", "slug", "topics", "unlocked_by"] - 1 tracks - ["php"]
  * ["uuid", "slug", "unlocked_by", "difficulty", "topics"] - 1 tracks - ["ecmascript"]
  * ["core", "difficulty", "slug", "topics", "unlocked-by", "uuid"] - 1 tracks - ["bash"]

Future directions to take this script:

* Some orderings with missing keys are potentially compatible with one another. For example, "A, B, D" and "A, C, D" are both compatible with "A, B, C, D". This script made no attempt to combine compatible orderings into one bucket.
* You may notice that tracks can and do appear in the exercises array more than once. No attempt was made to display how many times a track shows up.

> Now a lot of this I think depends on Go JSON handling if we are looking at configlet as the handler for these issues.

I shrug at this and ask why does it need to be configlet? `jq . < config.json` and `ruby -rjson -e 'puts JSON.pretty_generate(JSON.parse(ARGF.read))' < config.json` both format whitespace without changing order. Make this another reason why I my highest preference is only formatting whitespace: We get it for free without doing any work at all because other tools have already done it for us.

**Comment by petertseng** [comment]
> For example, "A, B, D" and "A, C, D" are both compatible with "A, B, C, D". This script made no attempt to combine compatible orderings into one bucket.

I would suggest a possible alternative to trying to find maximal orderings that many tracks are compatible with.

Instead, for every pair of keys that appears: count how many tracks say that A should go before B, versus how many tracks say that B should go before A.

**Comment by tleen** [comment]
> Please back up with data.

If you think there is a unified format then present it with data? That is how it gets fleshed out. I don't have a preference. You seem to?

Also `configlet` could do it or not, I just think if there is a de facto ordering configlet should respect it. It does not have to define it.

**Comment by tleen** [comment]
> Instead, for every pair of keys that appears: count how many tracks say that A should go before B, versus how many tracks say that B should go before A.

Or contributors that care can declare it here. I think its obvious sofar, perhaps it is not, but I think the proposed layout above is it, and I welcome maintainers to declare otherwise. `maximal` doesn't matter, the ordering should be in the format `ordering maintainers care about` and then the rest.

**Comment by kytrinyx** [comment]
I would prefer that the decision be made based on explicitly stated preferences of maintainers, rather than data coming from existing `config.json` files, which may or may not reflect intentional ordering.

**Comment by Insti** [comment]
* The current ordering can not be trusted as a proxy for ordering preference.
* I'm not sure that there would be strong objection to a fixed logical ordering.
* It seems silly to build tooling for a lot of flexibility if no-one really cares.

Does someone want to propose an 'official' fixed key ordering so we can see if there ARE any objections?

**Comment by petertseng** [comment]
> I'm not sure that there would be strong objection to a fixed logical ordering.

One possible objection is "hey, I don't have the kind of time to be the one who will be figuring out what the ordering is". True, everyone has limited time, but the response to this line of reasoning is "No no, we're not asking *that* of you. Instead, even if you don't want to decide the global ordering, *if* one were to exist *and* you want your preferences (whatever they are) to be respected in it, you are going to need to say them"

Other objections are along the lines of "Hey, these are just my personal preferences, who am I to force them on others?", or the reverse, "Who are you to force your ordering on me?" For such objections, supporters should come prepared with reasons why it is good to have the same ordering across multiple tracks.

> propose an 'official' fixed key ordering

for convenience (to avoid extra round-trips) I will definitely note that https://github.com/exercism/meta/issues/95#issuecomment-340291739 contains a proposal for keys in objects contained in `exercises`. That proposal should be evaluated.

To aid in crafting proposals, I will also note that a one-line addition to the script will output the union of all keys in current use: `puts "  * UNION: #{orderings.map(&:first).reduce([], :|)}"`.

And here they are:

* Path /
  * UNION: ["active", "exercises", "foregone", "language", "solution_pattern", "ignore_pattern", "ignored", "blurb", "test_pattern", "checklist_issue", "track_id", "deprecated", "gitter"]
  * I note that ignored doesn't need to exist anymore.
  * I note that deprecated has moved out of the top level and should therefore not exist at the top level anymore
* Path /exercises
  * UNION: ["core", "difficulty", "slug", "topics", "unlocked_by", "uuid", "deprecated", "unlocked-by"]
  * I will note that unlocked-by appears to be a mistake.

---

I am writing down my personal preferences so that one may understand how I will treat proposals. It seems good to have such things in writing. I advise readers not to consider this a proposal because it contains too much free-form thinking and is too wishy-washy to be one.

* Path /
  * language first; it seems important
  * track_id second; same
  * all _pattern should be contiguous in order
  * foregone should probably be immediately before exercises; they're related
  * exercises last, since it is a large list and I wouldn't want to have to make anyone scroll beyond it to hunt for anything coming after
* Path /exercises
  * uuid and slug first and second in some order, since they seem important to the exercise's identity
  * core and unlocked_by are both related to progression, so they should be contiguous in order. probably core before unlocked_by.
  * continuing with the trend of placing lists last, I'd probably put topics last
  * interesting, I haven't thought about where I want difficulty to go. It's currently immediately before topics because difficulty and topics were added together, but does that need to stay the case?

**Comment by petertseng** [comment]
EBWODP:

> I note that ignored need to exist anymore.

**Comment by petertseng** [comment]
EBWODP:

> I note that ignored need to exist anymore.

an obvious mistake on my part. ignored **doesn't** need to exist anymore.

**Comment by petertseng** [comment]
EBWODP:

> let's assume that one will exist sometime. 

The argument is more general if we word it "if one exists sometime"

**Comment by petertseng** [comment]
EBWODP:

> but the response to this line of reasoning is 

it's also "we're not asking that of you!"

**Comment by petertseng** [comment]
EBWODP:

> I advise readers not to consider this a proposal.

because it contains too much free-form thinking and is too wishy-washy to be one

**Comment by Insti** [comment]
Off-topic comments: 
`foregone` as a section should probably be deprecated and that indicator should be merged into the individual exercise object as `deprecated` has.
`core` seems to be determinable by the `unlocked_by` status?

Thanks to @petertseng for doing the work that made this post easy to write :heart:

---
On topic:


Human readable identifiers should come before machine friendly identifiers.

So here is a concrete proposal for a fixed key ordering:

    Path /
      track_id
      active
      language 
      blurb
      gitter
      checklist_issue
      ignore_pattern
      solution_pattern
      test_pattern
      foregone 
      exercises 

    Path /exercises
      slug
      uuid        
      core 
      unlocked_by
      difficulty
      topics

Thumb up if you're fine with this, thumb down if you want something different.

**Comment by NobbZ** [comment]
I haven't read through those longish posts. I currently can't focus enough for those.

In general, I do prefer a key-order which has identifiers first and lists last.

If we put UUID first or the slug, I do not care, as long as no mutable values appear above the actual slug or track-id.

The reason for putting lists last is, that large/long lists can span multiple screens and if there is a single entry after that list, it might hard to find that and to even recognize to which object it actually belongs (is it still hello-world or did I skip another exercise by accident and this belongs already to leap?)

**Comment by Insti** [comment]
Off-topic: I suspect `checklist_issue` is redundant and can be removed.
Origin: https://github.com/exercism/x-api/issues/109

**Comment by kytrinyx** [comment]
A few notes:

`foregone` is not the same as deprecated, but it would make sense to fold this into the `exercises` list. E.g.

```json
...
"exercises": [
    {
        "slug": "sparkles",
        "uuid": "87494fbe-a01e-4f02-a216-9e1ccf9314a5",
        "do-not-implement": true,
    }
]
...
```

`checklist_issue` is still necessary, unless we decide to not link to the track's starter issue. There was a time when we didn't have boilerplate for a new track (or starter issues), and so some old, unlaunched tracks have starter issues that are not the first issue. We could treat these tracks differently (e.g. not show them on the site) and thereby remove the need for that key in the config, but that seems like a different discussion from this one. For now, let's assume that we need to keep it.

There's another scripted cross-track cleanup task that I recalled that we'll need to do before we launch v2. We have a problem with the currently used UUID library in Configlet, so we'll need to regenerate/replace any invalid UUIDs.

**Comment by bdw429s** [comment]
It seems that we're fixated a bit on nailing down an order of JSON keys that every track maintainer can agree on, but can someone articulate for me the impetus for it?  As in, what is the tangible benifit of having the order of keys match across tracks?  I just re-scanned the entire conversation and here are really the two "problems" needing solved:

1. We need the ability to script JSON changes across all tracks in an automated fashion.
2. When No 1 happens, it should be minimally invasive to the individual track maintainers.  The specifics that have been complained about are:
    * Cluttered diffs that are hard to merge (in part, due to whitespace changes)
    * Keys being re-ordered to something different than before the scripted change.

Neither of those points require a globally-agreed upon order, so why are we still talking about one?  All they require is that the given order in a JSON file be _preserved_ when written back out for that track.  (which is something the CFML tooling is already doing)

**Comment by kytrinyx** [comment]
Consistent order is valuable to me in two ways (I can't speak for others).

1. A lot of people change a track's config.json file, and most people don't have strong preferences about order, which means that either
    1. the track maintainers end up playing the linter and asking people to change the order, or
    1. the config file itself becomes inconsistent (different exercise objects have keys ordered differently) making the file harder to consume as a human.
1. For maintainers who deal with multiple tracks (we have several) having the same order across all the tracks they deal with slightly reduces the mental overhead of working with the file.

I would imagine that if we decide on an order that is reasonable, it shouldn't be too hard to agree to use it. People feel very strongly that alphabetical is the _wrong_ order, but it doesn't seem like anyone would feel so strongly against a meaningful order that takes into account all the key points made above, even if it is not their _favorite_ order.

**Comment by petertseng** [comment]
This here comment only contains responses to remark that were deemed off-topic by their respective authors (okay, author, but WLOG).

> foregone as a section should probably be deprecated and that indicator should be merged into the individual exercise object as deprecated has.

I'm going to need a refresher on why `deprecated` was moved there. I guess https://github.com/exercism/discussions/issues/159#issuecomment-313878894 is it? I guess it makes sense to have deprecated exercises in the array since they still are visible on the site, etc.?

The same (visibility on the site) is not true of `foregone`, and I would much rather not have to use five lines (open brace, slug, uuid, do not implement true) to express what I used to be able to express with one. Why does a foregone exercise need a UUID anyway?

> core seems to be determinable by the unlocked_by status?

Depends on whether you allow exercises that are `core: false` and `unlocked_by: null`. These have been explicitly recommended in https://github.com/exercism/discussions/issues/175#issue-244918424 . As long as such exercises exist, it not possible to determine from `unlocked_by` alone the `core` status of any `unlocked_by: null` exercise. Please say the word if this recommendation is no longer recommended!

**Comment by Insti** [comment]
Thanks for the clarifications @kytrinyx & @petertseng :heart:

My off-topic comments were mainly to act as indicators that either
a) Maybe this is something that needs looking at.
or
a) The previous decisions are still valid and nothing needs to change.

I admit to not having done a lot of research into the comments as I made them.
But it seems like they are all cases of b) so require no further discussion in this thread.

**Comment by petertseng** [comment]
> the config file itself becomes inconsistent (different exercise objects have keys ordered differently) making the file harder to consume as a human.

Oh, I would dislike that. It would be good if I could have something that would make sure things are consistent.
  
I hope you don't see this comment as nitpicking, but rather as a way to recognise problems and think of solutions to them:

It occurs to me that a consistency checker need not have any specific order in mind; it only needs to check that all the orders it encounters are consistent with each other.

The upshot: If we do not agree on an ordering, one can at least write a consistency checker using that idea.

(With some small trickiness needed for `deprecated` vs non-`deprecated` exercises, but this can be dealt with)

> the track maintainers end up playing the linter and asking people to change the order
> [...]
> For maintainers who deal with multiple tracks (we have several) having the same order across all the tracks they deal with slightly reduces the mental overhead of working with the file

To contrast this to an alternative solution, one can conceive of a world where each maintainer writes a custom Travis CI check each enforcing the desired ordering of the track. It would work, but it would be some duplicate effort.

If there are those who would step up to contribute and avoid duplicate effort, that's commendable.

I definitely acknowledge that this is not a perfect analogue of the same situation of deduplication in other contexts (where I am guessing the adoption process is slower), but I don't think it's necessary for it to be a perfect analogue.

For the purpose of enabling scripted changes and what it means for any JSON formatter we choose to use, since I would accept the use of existing tools (`jq` or `ruby -rjson`) that require no additional implementation effort, it means I personally cannot rationally choose to take on any implementation effort, but this definitely shouldn't stop anyone else!

In the same vein, it does mean that if it is rational for others to take on the effort, and then they do, and then the result is acceptable, it would make sense to use it.

**Comment by cmccandless** [comment]
>To contrast this to an alternative solution, one can conceive of a world where each maintainer writes a custom Travis CI check each enforcing the desired ordering of the track. It would work, but it would be some duplicate effort.

I really like the idea of integrating config.json style checks into Travis CI. The Python track has had a lot of exercise implementation PRs this month, and many contributors (myself included) have had to be reminded on a few formatting conventions we've established there; this would eliminate a good amount of simple and repetitive code review needed.

**Comment by Insti** [comment]
@cmccandless I suggest raising an issue in the Python track for this.
Many tracks already have this set up, but it is something that needs to be done at a track level.

**Comment by Stargator** [comment]
The Dart track just follows the configlet format output. So we would like any changes to come from there.

I would feel it would cause confusion if we deviated from that.

**Comment by kytrinyx** [comment]
I've read through this thread again from start to finish, and these are my observations:

* Some maintainers care about consistency, some don't.
* Some maintainers care about order, some don't.

Consistency appears to be about two things:
* reducing mental overhead when reading or changing the config file.
* making automation easier.

Order appears to be about two things that both boil down to reducing mental overhead:
* consistency
* meaning

Here are the things we want to optimize for:

* Minimize work for maintainers.
    * Maintainers should not have to mentally perform linting on contributions.
    * Maintainers should not have to write their own tooling to deal with the config.
    * Maintainers should be able to look at a pull request and quickly know what meaningful change has been made.
* Optimize for human readability.
    * Order should be roughly by _importance_.
    * Order of elements for exercise objects should be consistent within a track.
    * Long entries (lists or long bits of text) should be at the bottom.
* Optimize for scriptability of `config.json` (and perhaps less importantly `maintainers.json`).
    * It should not make unrelated changes to whitespace.
    * It should not reorder elements.

There are two approaches that have gotten a lot of attention above:

* Only reformat whitespace.
* Decide on an order and enforce it (along with updating whitespace).

It sounds like the biggest arguments for having a formatter that only touches whitespace are about:
* freedom of expression (allowing maintainers to decide, individually, what they think is important)
* not abusing JSON as a tool (JSON is unordered, therefore our config should be unordered)
* simplicity (there's no need to serialize/deserialize the JSON to make changes)

The trade-off of allowing maintainers to decide, individually, what they think is important, is that maintainers have to decide, individually, what is important. While feelings are clearly very strong about the subject, I don't actually think that making 40+ teams of maintainers decide ordering is a very good use of time and energy.

There have been half-serious suggestions about choosing a different file format than JSON for this. I don't see any strong arguments for switching, though, so I'm not currently considering that. I'm perfectly comfortable abusing JSON for this purpose.

In terms of serialization, updating only whitespace is simpler for reformatting, but adds brittleness and complexity when I need to change the JSON data. I find that serializing/deserializing is simpler,  since that allows me to work in a native data type, rather than trying to munge a string and hope it turns out to be valid JSON.

So. I'm making the call.

**We will pick an acceptable order and update configlet's `fmt` command to use it.**

_Acceptable_ means just that. Not the ultimate best possible perfect amazingest wishful thinking order, since I think that not everyone could agree on that one. I'm pretty sure that we can come up with something that is perfectly reasonable, and therefore acceptable.


For the maintainers config:

* Username should be first
* Then alumnus
* Then show on website
* Then it doesn't matter much.
* Bio last (because it's potentially the longest)

Here are the characteristics that have come up that seem important for the `config.json` file:

* Put the "exercises" list at the bottom, and all the short config values at the top.
* Put "language" at or near the top.
* Within exercises
    * Put UUID and slug at the top. It probably doesn't matter which of those are first, but I'd lean towards the slug because that is the human-readable identifier.
    * Topics last, because it's a list.
    * Core and unlocked by should be together, because they belong together.

I've started with @Insti's suggestion above, swapping around `active` and `language` because `track_id` and `language` feel like they belong together.

* Path `/`
    1. track_id
    1. language 
    1. active
    1. blurb
    1. gitter
    1. checklist_issue
    1. ignore_pattern
    1. solution_pattern
    1. test_pattern
    1. foregone 
    1. exercises 
* Path `/exercises`
    1. slug
    1. uuid        
    1. core 
    1. unlocked_by
    1. difficulty
    1. topics

I think this is perfectly reasonable, but I honestly don't care very hard about the specifics as long as they're meaningful enough.

For exercises we also have the `deprecated`, would could go last, with the note that when an exercise is deprecated, we delete all keys except `uuid`, `slug`, and `deprecated`.

If anyone has any arguments for a different ordering, please make your arguments now.

**Comment by kytrinyx** [comment]
Latest update: a couple of tweaks have been requested to the configlet fmt command:

exercism/configlet#129
exercism/configlet#128

**Comment by kytrinyx** [comment]
Update: I've reworked the fmt command for configlet. I am waiting for a code review, and expect to be able to cut a release this week so we can close this issue out.

**Comment by kytrinyx** [comment]
I've published a new release of configlet: https://github.com/exercism/configlet/releases/tag/v3.9.1

**Comment by kytrinyx** [comment]
@laurafeier @paulgoetze - would either of you happen to have the script handy from ROSSConf? If so, would you be interested in taking another stab at submitting the result of `configlet fmt` to the tracks? (I'd recommend running a "canary" first against a small number of tracks, if so).

**Comment by rpottsoh** [comment]
Already ran `fmt` it on Delphi.  Looks good.  👍

**Comment by kytrinyx** [comment]
Sweet, thanks for taking a look at that, @rpottsoh! I can't wait to close this issue :-)

**Comment by cmccandless** [comment]
Will there be an automated PR to each track as needed, or should track maintainers go ahead and run the new release against their respective tracks on their own?

**Comment by rpottsoh** [comment]
I think @kytrinyx is planning to submit PRs for all the tracks via a script.  I just happened to notice that a new release was available and jumped in and ran it manually.  It looks like she is just trying to track down a (the) script to do this.  I am just basing this on her [earlier comment](https://github.com/exercism/meta/issues/95#issuecomment-394209150).

**Comment by laurafeier** [comment]
@kytrinyx I'm away this week but if someone wants to give it a try we used this [script](https://gist.github.com/laurafeier/ef9b0fc37425d9b4cf3ec47c85b7fb60) to create MR for each project. The forking part is missing but you guys can change it a bit to make the magic happen.

**Comment by kytrinyx** [comment]
I've submitted canary PRs to a handful of tracks.
